crates/db_schema/src/impls/actor_language.rs

```diff
@@ -0,0 +1,437 @@
+use crate::{
+  diesel::JoinOnDsl,
+  newtypes::{CommunityId, LanguageId, LocalUserId, SiteId},
+  source::{actor_language::*, language::Language},
+};
+use diesel::{
+  delete,
+  dsl::*,
+  insert_into,
+  result::Error,
+  select,
+  ExpressionMethods,
+  PgConnection,
+  QueryDsl,
+  RunQueryDsl,
+};
+use lemmy_utils::error::LemmyError;
+
+impl LocalUserLanguage {
+  pub fn read(
+    conn: &mut PgConnection,
+    for_local_user_id: LocalUserId,
+  ) -> Result<Vec<LanguageId>, Error> {
+    use crate::schema::local_user_language::dsl::*;
+
+    local_user_language
+      .filter(local_user_id.eq(for_local_user_id))
+      .select(language_id)
+      .get_results(conn)
+  }
+
+  /// Update the user's languages.
+  ///
+  /// If no language_id vector is given, it will show all languages
+  pub fn update(
+    conn: &mut PgConnection,
+    language_ids: Vec<LanguageId>,
+    for_local_user_id: LocalUserId,
+  ) -> Result<(), Error> {
+    conn.build_transaction().read_write().run(|conn| {
+      use crate::schema::local_user_language::dsl::*;
+      // Clear the current user languages
+      delete(local_user_language.filter(local_user_id.eq(for_local_user_id))).execute(conn)?;
+
+      let lang_ids = update_languages(conn, language_ids)?;
+      for l in lang_ids {
+        let form = LocalUserLanguageForm {
+          local_user_id: for_local_user_id,
+          language_id: l,
+        };
+        insert_into(local_user_language)
+          .values(form)
+          .get_result::<Self>(conn)?;
+      }
+      Ok(())
+    })
+  }
+}
+
+impl SiteLanguage {
+  pub fn read_local(conn: &mut PgConnection) -> Result<Vec<LanguageId>, Error> {
+    use crate::schema::{site, site_language::dsl::*};
+    let subquery = crate::schema::site::dsl::site
+      .order_by(site::id)
+      .select(site::id)
+      .limit(1)
+      .into_boxed();
+    site_language
+      .filter(site_id.eq_any(subquery))
```

Comment:
Put a TODO here because this subquery is bad. This should go away once we add a site:local column.

Reply:
done